File: mirea/cources/2026-2027/aii-july/06/Part-02-Recursive-generation-basics/examples/example-04.cpp
// Пример: все подмножества множества {1, ..., n}.
// На каждом шаге рекурсия решает, брать текущий элемент или нет.

#include <iostream>
#include <vector>

using namespace std;

vector<int> current;

// Печатаем одно готовое подмножество в виде {1, 3, 5}.
void print_subset() {
    cout << '{';
    for (int i = 0; i < (int)current.size(); i++) {
        if (i) cout << ", ";
        cout << current[i];
    }
    cout << "}\n";
}

// Для каждого value решаем: брать его в подмножество или нет.
void dfs(int value, int n) {
    if (value > n) {
        // Построили одно полное подмножество.
        print_subset();
        return;
    }

    // Ветка 1: не берём value.
    dfs(value + 1, n);

    // Ветка 2: берём value.
    current.push_back(value);
    dfs(value + 1, n);
    current.pop_back();
}

int main() {
    int n;
    // Считываем n и запускаем перебор с первого числа.
    cin >> n;

    dfs(1, n);
    return 0;
}

File: mirea/cources/2026-2027/aii-july/06/Part-02-Recursive-generation-basics/examples/example-04.py
# Пример: все подмножества множества {1, ..., n}.
# На каждом шаге рекурсия решает, брать текущий элемент или нет.

n = int(input())
current = []


# Печатаем одно готовое подмножество в виде {1, 3, 5}.
def print_subset() -> None:
	print('{' + ', '.join(map(str, current)) + '}')


# Для каждого value решаем: брать его в подмножество или нет.
def dfs(value: int) -> None:
	if value > n:
		# Построили одно полное подмножество.
		print_subset()
		return

	# Ветка 1: не берём value.
	dfs(value + 1)

	# Ветка 2: берём value.
	current.append(value)
	dfs(value + 1)
	current.pop()


# Запускаем перебор с первого числа.
dfs(1)